src/Controller/DisplayWaitingForPaymentPage.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+declare(strict_types=1);
+
+namespace CommerceWeavers\SyliusTpayPlugin\Controller;
+
+use Payum\Core\Payum;
+use Sylius\Bundle\PayumBundle\Factory\ResolveNextRouteFactoryInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\RouterInterface;
+use Twig\Environment as Twig;
+
+final class DisplayWaitingForPaymentPage
+{
+    public function __construct(
+        private readonly Payum $payum,
+        private readonly RouterInterface $router,
+        private readonly ResolveNextRouteFactoryInterface $resolveNextRouteFactory,
+        private readonly Twig $twig,
+        private readonly int $refreshInterval,
+    ) {
+    }
+
+    public function __invoke(Request $request): Response
+    {
+        $token = $this->payum->getHttpRequestVerifier()->verify($request);
+
+        $resolveNextRoute = $this->resolveNextRouteFactory->createNewWithModel($token);
+
+        $this->payum->getGateway($token->getGatewayName())->execute($resolveNextRoute);
+
+        if (
+            is_string($resolveNextRoute->getRouteName()) &&
+            $resolveNextRoute->getRouteName() !== 'commerce_weavers_tpay_waiting_for_payment'
```

Comment:
once we have route class, would be nice to use it here 
```suggestion
            $resolveNextRoute->getRouteName() !== 'Route::SHOP_WAITING_FOR_PAYMENT'
```